File: jscomp/test/test_string_map.js
'use strict';

var $$Map = require("melange/lib/js/map.js");
var Caml = require("melange/lib/js/caml.js");
var Curry = require("melange/lib/js/curry.js");

var compare = Caml.caml_string_compare;

var StringMap = $$Map.Make({
      compare: compare
    });

function timing(label, f) {
  console.time(label);
  Curry._1(f, undefined);
  console.timeEnd(label);
}

function assertion_test(param) {
  var m = {
    contents: StringMap.empty
  };
  timing("building", (function (param) {
          for(var i = 0; i <= 1000000; ++i){
            m.contents = Curry._3(StringMap.add, String(i), String(i), m.contents);
          }
        }));
  timing("querying", (function (param) {
          for(var i = 0; i <= 1000000; ++i){
            Curry._2(StringMap.find, String(i), m.contents);
          }
        }));
}

exports.assertion_test = assertion_test;
/* StringMap Not a pure module */

File: jscomp/test/test_string_map.ml
include (struct
module StringMap = Map.Make(struct type t = string
  let compare
      (x : string) y = compare x y end)
external time : string -> unit = "console.time"
external timeEnd : string -> unit = "console.timeEnd"

let timing label f =
  begin
    time label;
    f ();
    timeEnd label;
  end

let assertion_test () =
  let m = ref StringMap.empty in
  let count = 1000000 in
  timing "building" @@ (fun _ ->

  for i = 0 to count do
    m := StringMap.add (string_of_int i) (string_of_int i) !m
  done);
  timing "querying" @@ (fun _ ->
  for i = 0 to count  do
    ignore (StringMap.find (string_of_int i) !m )
  done)


end : sig val assertion_test : unit -> unit end)
